src/Common/src/System/Collections/Generic/ArrayBuilder.cs

```diff
@@ -80,53 +68,66 @@ public void AddRange(T[] items)
             }
         }
 
-        public T[] AsArray() => _array ?? Array.Empty<T>();
-
-        public T[] AsArray(out int count)
-        {
-            count = _count;
-            return AsArray();
-        }
-
-        public T[] AsOrToArray()
-        {
-            return _count == Capacity ? AsArray() : ToArray();
-        }
-
         public Enumerator GetEnumerator()
         {
             return new Enumerator(_array, _count);
         }
 
         public T[] ToArray()
         {
-            if (_count == 0)
+#if DEBUG
+            // Try to prevent callers from using the ArrayBuilder after ToArray.
+            _count = -1;
```

Comment:
Just before `return _array`? So then people could still use the builder if they call `ToArray` on an empty builder?